### ISSUE TITLE
Mishandles `self` argument of wrapped method 

### ISSUE THREAD

**Issue by rparini** [microsoft/pyright#4843]
**Describe the bug**
Pyright does not recognise that an instance of a class is implicitly passed to the `self` argument of a class method decorated with `@functools.wraps`.

**To Reproduce**
With pyright `1.1.300` and Python 3.9.13
```python
import functools

class A:
    def meth(self, x):
        return x

class B:
    @functools.wraps(A.meth)
    def meth(self, x):
        return x + 1

b = B()
y = b.meth(x)
```
Results in two errors
```
pyright 1.1.300
/Users/rparini/Documents/examples/wraps_pyright.py
  /Users/rparini/Documents/examples/wraps_pyright.py:13:5 - error: Argument missing for parameter "x" (reportGeneralTypeIssues)
  /Users/rparini/Documents/examples/wraps_pyright.py:13:12 - error: "x" is not defined (reportUndefinedVariable)
2 errors, 0 warnings, 0 informations 
```
It seems to expect `B.meth(b, x)` rather than `b.meth(x)`

**Expected behavior**
I believe pyright should report no errors in the above. Pyright `1.1.285` does not report any errors but `1.1.298` and onwards does so I suspect this is an issue with pyright, rather than typeshed.

**VS Code extension or command-line**
I find an the error reported in this case with both the command line pyright 1.1.300 and Pylance 
> Pylance language server 2023.3.20 (pyright 93a4ef87) starting

**Comment by erictraut** [comment]
The change is due to a recent change in the typeshed stub file `functools.pyi`. Pyright doesn't have any special-case handling for `functools.wraps`. The change in typeshed was made to improve the handling of `functools.wraps`, but it does introduce regressions in some cases like the one you've pointed out here. You can see the new type definition if you right click on `wraps` and choose `Go To Declaration`. It makes use of the `ParamSpec` feature of the type system to capture the signature of the wrapped function or method.

I don't see an obvious way to make this case work through modification of the typeshed declaration without breaking other (more common) use cases of `wraps`. The only other solution I can think of is to completely ignoring the typeshed definition and specialcase all of the `functools.wraps` logic in the type checker. That's something we generally try not to do because it is fragile and increases the maintenance burden of the type checker.

I'm curious what you're trying to do with this code. I don't think I've seen `functools.wraps` used in this manner before. Are you simply trying to copy the `__doc__`, `__name__`, `__annoations__`, etc. fields from `A.meth` to `B.meth` but otherwise call through directly to `B.meth`? It doesn't look like you're actually creating a wrapper function here.

**Comment by erictraut** [comment]
Here's a potential workaround:

```python
class B:
    def meth(self, x: int) -> int:
        return x + 1

    # Copy docstring from A.meth
    functools.update_wrapper(meth, A.meth)
```
